gh-75666: Fix Misc.unbind behaviour when funcid is given

GiovaLomba opened the pull request:
<!-- gh-issue-number: gh-75666 -->
* Issue: gh-75666
<!-- /gh-issue-number -->

GiovaLomba committed:
Fix: https://bugs.python.org/issue31485

Allows Misc.unbind to delete only the bound callback
having the given funcid. See the issue above for more.

GiovaLomba committed:
Merge pull request #1 from GiovaLomba/issue31485

bpo-31485: Fix Misc.unbind behaviour when funcid is given

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

terryjreedy commented on Lib/tkinter/__init__.py:
```diff
@@ -1385,11 +1385,14 @@ def bind(self, sequence=None, func=None, add=None):
+        FUNCID is given, delete the command also."""
+        bound = ''
         if funcid:
             self.deletecommand(funcid)
```
The test failure is here, `_tkinter.TclError: can't delete Tcl command`.  In the absence of more specific information, I interpret this as tk, on Linux, refusing to delete a command that is in use.  Try moving the deletion back to after the rebinding, as it originally was.  It is OK to have two `if funcid` clauses.

GiovaLomba replied:
Should this work? I've just given it a try on Ubuntu and tests worked.
```
    def unbind(self, sequence, funcid=None):
        """Unbind for this widget the event SEQUENCE. if
        FUNCID is given, delete the command also."""
        funcs = self.tk.call('bind', self._w, sequence, None)
        self.tk.call('bind', self._w, sequence, '')
        if funcid:
            self.deletecommand(funcid)
            keep = '\n'.join(f for f in funcs.split('\n') if not f.startswith(f'if {{"[{funcid}'))
            self.tk.call('bind', self._w, sequence, keep)
```
Where should the test be written? Would the test just check that after `unbind` with given `funcid`, other bound function are there and the given one is gone?

GiovaLomba replied:
Hey @terryjreedy r you still there? I've done everything you asked except for the new test case, for which I asked you about. As you see the code now passes all tests and could be merged. I mark this as resolved.

terryjreedy replied:
The Travis test still fails and the revision is not what I asked for, apparently not clearly enough.  A second problem is that the new docstring is what I suggested on the issue better described the current behavior, not the desired behavior.  I have revised the docstring to what I believe Serhiy wanted and the the code to match.  If CI passes, we can consider a new test.

terryjreedy commented on Lib/tkinter/__init__.py:
```diff
@@ -1385,11 +1385,14 @@ def bind(self, sequence=None, func=None, add=None):
-        self.tk.call('bind', self._w, sequence, '')
+        """Unbind for this widget the event SEQUENCE. if
+        FUNCID is given, delete the command also."""
+        bound = ''
```
`keep` or `rebind` strike me as better names, but follow Serhiy if he disagrees.

GiovaLomba replied:
I'm using `keep`.

terryjreedy commented on Lib/tkinter/__init__.py:
```diff
@@ -1385,11 +1385,14 @@ def bind(self, sequence=None, func=None, add=None):
         if funcid:
             self.deletecommand(funcid)
+            funcs = self.tk.call('bind', self._w, sequence, None).split('\n')
+            bound = '\n'.join([f for f in funcs if not f.startswith(f'if {{"[{funcid}')])
```
`join` does not need the intermediate list and the second iteration.  Remove '[' and ']' to leave a generator expression (comprehension).

I don't understand `f.startswith(f'if {{"[{funcid}')`, ie, why the func corresponding to funcid should necessarily be a string starting with something like `if {"[<funcid>`.  Does this have something to do with how tcl wraps python functions?  If so, should we assume that the bound function is such?

GiovaLomba replied:
You'right for the generator. 
Neither do I. After having done several tests to check how callbacks are bound, it came out that they always start with that pattern, so I assumed this is "how tcl wraps python functions".

terryjreedy commented on Misc/NEWS.d/next/Library/2020-01-11-22-09-29.bpo-31485.DSvjsY.rst:
```diff
@@ -0,0 +1 @@
+On the tkinter library, the widget.unbind(<Sequence>, funcid) call unbinds all bindings for <Sequence>, while documentation states that when funcid is given all other bindings remains untuched. Apparently this behaviour is not offered by tkinter itself and, in case of non-empty funcid argument value be given, should be obtained by, at first, unbinding all bindings and then bind again all the ones ...
```
This is one line of about 500 chars.  I think we usually comment with line breaks, but I don't know if it makes much difference.  In any case, this seems like more explanation than needed by users, who are the audience for news items.

The technical details can be appropriate for the commit message.  You can edit the initial message on the PR to include a suggested commit message (about the bpo link).

terryjreedy reviewed (changes requested):
I think a new test is needed to verify that the revision does what we want and expect it to.

GiovaLomba committed:
Fix: tests on Ubuntu bionic

Fix `_tkinter.TclError: can't delete Tcl command` on Ubuntu bionic.
Change `bound` to `keep` as suggested.

GiovaLomba committed:
Update 2020-01-11-22-09-29.bpo-31485.DSvjsY.rst

terryjreedy committed:
Update __init__.py

Revise docstring to match behavior specified by Serhiy.  Rearrange code to match.

terryjreedy committed:
Fix whitespace, delete extra '.'

terryjreedy committed:
Leave no-funcid call as it was.

GiovaLomba commented on Lib/tkinter/__init__.py:
```diff
@@ -1385,14 +1385,20 @@ def bind(self, sequence=None, func=None, add=None):
+        and also delete that command.
+        """
+        if funcid is None:
+            self.tk.call('bind', self._w, sequence, None)
```
Hey @terryjreedy `None` as third argument to `bind` does not unbind nothing! Take a look at the doc [https://www.tcl.tk/man/tcl8.6/TkCmd/bind.htm](https://www.tcl.tk/man/tcl8.6/TkCmd/bind.htm).

terryjreedy replied:
"If script is an empty string then the current binding for sequence is destroyed, leaving sequence unbound.", which is the intent of the first call.  An empty string is returned ("In all of the cases where a script argument is provided, bind returns an empty string.")

"If sequence is specified without a script, then the script currently bound to sequence is returned, or an empty string is returned if there is no binding for sequence. "  This is the intent of the second call.  The explicit None is not needed since it is the default.

The code for _bind treats '' and None very differently.

GiovaLomba replied:
@terryjreedy I don't have a lot of spare time these days. I saw you've now correctly removed the `None`  with the empty `str`, as the documentation was pointing, and I'm happy with it. Eventually now this almost 2yo child bug will be fixed in the coming releases ;-)

How can we request the backport of this PR to 3.7 and 3.8?

terryjreedy replied:
With labels.  I consider this a bugfix, but as Serhiy is the main tkinter maintainer, will let him decide on backporting.

GiovaLomba replied:
Ok! Thank for your answer and for your time :)

terryjreedy committed:
Remove redundant split

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2020-01-11-22-09-29.bpo-31485.DSvjsY.rst:
```diff
@@ -0,0 +1 @@
+https://github.com/python/cpython/pull/17954#
```
Please add a short description, not a link.

There will be a link to the corresponded issue on the tracker in any case.

serhiy-storchaka reviewed (commented):
Looks correct to me, but it would be nice to add tests.
